Omitting `Clone` bound on `&impl Trait` leads to correct but unhelpful error

### Code

```Rust
#[derive(Clone)]
struct ThingThatDoesAThing {}

trait DoesAThing {}

impl DoesAThing for ThingThatDoesAThing {}

fn clones_impl_ref_inline(thing: &impl DoesAThing) {
    drops_impl_owned(thing.clone());
}

fn drops_impl_owned(_thing: impl DoesAThing) { }
```


### Current output

```Shell
error[E0277]: the trait bound `&impl DoesAThing: DoesAThing` is not satisfied
  --> src/lib.rs:9:22
   |
9  |     drops_impl_owned(thing.clone());
   |     ---------------- ^^^^^^^^^^^^^ the trait `DoesAThing` is not implemented for `&impl DoesAThing`
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `DoesAThing` is implemented for `ThingThatDoesAThing`
note: required by a bound in `drops_impl_owned`
  --> src/lib.rs:12:34
   |
12 | fn drops_impl_owned(_thing: impl DoesAThing) { }
   |                                  ^^^^^^^^^^ required by this bound in `drops_impl_owned`
```


### Desired output

```Shell
error[E0277]: the trait bound `&impl DoesAThing: DoesAThing` is not satisfied
  --> src/lib.rs:12:22
   |
 9 |     drops_impl_owned(thing.clone());
   |     ---------------- ^^^^^^^^^^^^^ the trait `DoesAThing` is not implemented for `&impl DoesAThing`
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `DoesAThing` is implemented for `ThingThatDoesAThing`
help: the type `&impl DoesAThing` does not implement `Clone`, so calling `clone` on `&impl DoesAThing` copies the reference, which does not do anything. Consider further restricting this bound
   |
 8 | fn clones_impl_ref_inline(thing: &impl DoesAThing + Clone) {
   |                                                      +++++++
note: required by a bound in `drops_impl_owned`
  --> src/lib.rs:33:34
   |
12 | fn drops_impl_owned(_thing: impl DoesAThing) { }
   |                                  ^^^^^^^^^^ required by this bound in `drops_impl_owned`
```


### Rationale and extra context

The current error message does not guide the programmer to the correct solution, which is to add a `Clone` bound to the type constraints of the generic parameter.

### Other cases


Moving the inline call to `thing.clone()` onto its own line improves the error, but it's still misleading:

```rust
fn clones_impl_ref_statement(thing: &impl DoesAThing) {
    let thing_clone = thing.clone();
    drops_impl_owned(thing_clone);
}
```
```
error[E0277]: the trait bound `&impl DoesAThing: DoesAThing` is not satisfied
  --> src/lib.rs:18:22
   |
18 |     drops_impl_owned(thing_clone);
   |                      ^^^^^^^^^^^ the trait `DoesAThing` is not implemented for `&impl DoesAThing`
   |
help: consider further restricting this bound
   |
16 | fn clones_impl_ref_statement(thing: &impl DoesAThing + Clone) {
   |                                                      +++++++
help: consider using clone here
   |
18 |     drops_impl_owned(thing_clone.clone());
   |                                 ++++++++
```

Removing the call to `drops_impl_owned()` allows compilation to proceed to the warning stage, which generates a warning that actually explains the problem (incorporated into Desired output, above):
```rust
fn clones_and_drops_impl_ref(thing: &impl DoesAThing) {
    let _ = thing.clone();
}
```
```
warning: call to `.clone()` on a reference in this situation does nothing
  --> src/lib.rs:23:18
   |
23 |     let _ = thing.clone();
   |                  ^^^^^^^^ help: remove this redundant call
   |
   = note: the type `impl DoesAThing` does not implement `Clone`, so calling `clone` on `&impl DoesAThing` copies the reference, which does not do anything and can be removed
   = note: `#[warn(noop_method_call)]` on by default
```


### Rust Version

```Shell
rustc 1.78.0-nightly (397937d81 2024-02-22)
binary: rustc
commit-hash: 397937d812852f9bbeb671005cb399dbcb357cde
commit-date: 2024-02-22
host: aarch64-apple-darwin
release: 1.78.0-nightly
LLVM version: 18.1.0
```


### Anything else?

I've set up a minimal reproducer repo for each of the above cases at [https://github.com/overhacked/impl-ref-reproducer](https://github.com/overhacked/impl-ref-reproducer)

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
@@ -987,15 +987,44 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
             else {
                 return false;
             };
-            let arg_node = self.tcx.hir_node(*arg_hir_id);
-            let Node::Expr(Expr { kind: hir::ExprKind::Path(_), .. }) = arg_node else {
-                return false;
-            };
 
             let clone_trait = self.tcx.require_lang_item(LangItem::Clone, None);
             let has_clone = |ty| {
                 self.type_implements_trait(clone_trait, [ty], obligation.param_env)
                     .must_apply_modulo_regions()
+            };
+
+            let existing_clone_call = match self.tcx.hir_node(*arg_hir_id) {
+                // It's just a variable. Propose cloning it.
+                Node::Expr(Expr { kind: hir::ExprKind::Path(_), .. }) => None,
+                // It's already a call to `clone()`. We might be able to suggest
+                // adding a `+ Clone` bound, though.
+                Node::Expr(Expr {
+                    kind:
+                        hir::ExprKind::MethodCall(
+                            hir::PathSegment { ident, .. },
+                            _receiver,
+                            &[],
+                            call_span,
+                        ),
+                    hir_id,
+                    ..
+                }) if ident.name == sym::clone
+                    && !call_span.from_expansion()
+                    && !has_clone(*inner_ty) =>
+                {
+                    // We only care about method calls corresponding to the real `Clone` trait.
+                    let Some(typeck_results) = self.typeck_results.as_ref() else { return false };
+                    let Some((DefKind::AssocFn, did)) = typeck_results.type_dependent_def(*hir_id)
+                    else {
+                        return false;
+                    };
+                    if self.tcx.trait_of_item(did) != Some(clone_trait) {
+                        return false;
+                    }
+                    Some(ident.span)
+                }
+                _ => return false,
             };
 
             let new_obligation = self.mk_trait_obligation_with_new_self_ty(
@@ -1015,12 +1044,23 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                         None,
                     );
                 }
-                err.span_suggestion_verbose(
-                    obligation.cause.span.shrink_to_hi(),
-                    "consider using clone here",
-                    ".clone()".to_string(),
-                    Applicability::MaybeIncorrect,
-                );
+                if let Some(existing_clone_call) = existing_clone_call {
+                    err.span_note(
+                        existing_clone_call,
+                        format!(
+                            "this `clone()` copies the reference, \
+                            which does not do anything, \
+                            because `{inner_ty}` does not implement `Clone`"
+                        ),
+                    );
+                } else {
+                    err.span_suggestion_verbose(
+                        obligation.cause.span.shrink_to_hi(),
+                        "consider using clone here",
+                        ".clone()".to_string(),
+                        Applicability::MaybeIncorrect,
+                    );
+                }
                 return true;
             }
             false

--- a/tests/ui/suggestions/clone-bounds-121524.rs
+++ b/tests/ui/suggestions/clone-bounds-121524.rs
@@ -1,0 +1,19 @@
+#[derive(Clone)]
+struct ThingThatDoesAThing;
+
+trait DoesAThing {}
+
+impl DoesAThing for ThingThatDoesAThing {}
+
+fn clones_impl_ref_inline(thing: &impl DoesAThing) {
+    //~^ HELP consider further restricting this bound
+    drops_impl_owned(thing.clone()); //~ ERROR E0277
+    //~^ NOTE copies the reference
+    //~| NOTE the trait `DoesAThing` is not implemented for `&impl DoesAThing`
+}
+
+fn drops_impl_owned(_thing: impl DoesAThing) { }
+
+fn main() {
+    clones_impl_ref_inline(&ThingThatDoesAThing);
+}

--- a/tests/ui/suggestions/clone-bounds-121524.stderr
+++ b/tests/ui/suggestions/clone-bounds-121524.stderr
@@ -1,0 +1,19 @@
+error[E0277]: the trait bound `&impl DoesAThing: DoesAThing` is not satisfied
+  --> $DIR/clone-bounds-121524.rs:10:22
+   |
+LL |     drops_impl_owned(thing.clone());
+   |                      ^^^^^^^^^^^^^ the trait `DoesAThing` is not implemented for `&impl DoesAThing`
+   |
+note: this `clone()` copies the reference, which does not do anything, because `impl DoesAThing` does not implement `Clone`
+  --> $DIR/clone-bounds-121524.rs:10:28
+   |
+LL |     drops_impl_owned(thing.clone());
+   |                            ^^^^^
+help: consider further restricting this bound
+   |
+LL | fn clones_impl_ref_inline(thing: &impl DoesAThing + Clone) {
+   |                                                   +++++++
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0277`.

